src/main/java/net/minecraftforge/fml/common/MissingModsException.java

```diff
@@ -19,71 +19,117 @@
         {
-            missingModsVersions.add(Pair.of(artifactVersion, null));
+            missingModsInfos.add(new MissingModInfo(artifactVersion, null, true));
         }
         this.modName = name;
     }
 
-    public MissingModsException(List<Pair<ArtifactVersion, ArtifactVersion>> missingModsVersions, String id, String name)
+    @Override
+    public String getMessage()
     {
-        super(String.format("Mod %s (%s) requires %s", id, name, missingModsVersions));
-        this.missingMods = new HashSet<>();
-        for (Pair<ArtifactVersion, ArtifactVersion> missingModsVersion : missingModsVersions)
-        {
-            this.missingMods.add(missingModsVersion.getKey());
-        }
-        this.missingModsVersions = missingModsVersions;
-        this.modName = name;
+        Set<ArtifactVersion> missingMods = missingModsInfos.stream().map(MissingModInfo::getNeededVersion).collect(Collectors.toSet());
+        return String.format("Mod %s (%s) requires %s", id, name, missingMods);
+    }
+
+    public void addMissingMod(ArtifactVersion needVersion, @Nullable ArtifactVersion haveVersion, boolean required)
+    {
+        MissingModInfo missingModInfo = new MissingModInfo(needVersion, haveVersion, required);
+        this.missingModsInfos.add(missingModInfo);
+        this.missingMods.add(needVersion);
     }
 
     public String getModName()
     {
         return modName;
     }
 
-    public List<Pair<ArtifactVersion, ArtifactVersion>> getMissingModsVersions()
+    public List<MissingModInfo> getMissingModInfos()
     {
-        return missingModsVersions;
+        return Collections.unmodifiableList(this.missingModsInfos);
     }
 
     @Override
     protected void printStackTrace(WrappedPrintStream stream)
     {
         stream.println("Missing Mods:");
-        for (Pair<ArtifactVersion, ArtifactVersion> entry : missingModsVersions)
+        for (MissingModInfo info : this.missingModsInfos)
         {
-            ArtifactVersion needVersion = entry.getKey();
-            ArtifactVersion haveVersion = entry.getValue();
+            ArtifactVersion needVersion = info.getNeededVersion();
+            ArtifactVersion haveVersion = info.getHaveVersion();
             String haveString = haveVersion != null ? haveVersion.getVersionString() : "null";
-            stream.println(String.format("\t%s : need %s: have %s", needVersion.getLabel(), needVersion.getRangeString(), haveString));
+            stream.println(String.format("\t%s : need %s: have %s", needVersion.getVersionString(), needVersion.getRangeString(), haveString));
         }
         stream.println("");
     }
+
+    public static class MissingModInfo
+    {
+        private final ArtifactVersion neededVersion;
+        @Nullable
+        private final ArtifactVersion haveVersion;
+        private final boolean required;
+
+        private MissingModInfo(ArtifactVersion neededVersion, @Nullable ArtifactVersion haveVersion, boolean required)
```

Comment:
I suggest names of `requiredVersion` and `currentVersion`; they have a formal tone.

Reply:
I agree those sound better, I'll change it now.

Reply:
well, actually I want something for `requiredVersion` that doesn't have `required` in the name, since there is a boolean for required too.

Reply:
Going to go with `acceptedVersion` and `currentVersion`